File: movieland-backend/src/main/java/com/phil/movieland/rest/service/MovieShowService.java
package com.phil.movieland.rest.service;

import com.phil.movieland.data.entity.Movie;
import com.phil.movieland.data.entity.MovieShow;
import com.phil.movieland.data.repository.MovieShowRepository;
import com.phil.movieland.utils.DateUtils;
import com.phil.movieland.utils.Utils;
import java.util.Calendar;
import java.util.Comparator;
import java.util.Date;
import java.util.List;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Service to interface movieShowRepository
 */
@Service
@Transactional
public class MovieShowService {

    private final MovieShowRepository movieShowRepository;

    private Logger log = LoggerFactory.getLogger(MovieShowService.class);

    @Autowired
    public MovieShowService(MovieShowRepository movieShowRepository) {
        this.movieShowRepository = movieShowRepository;
    }

    public List<MovieShow> getShowsForMovie(Movie movie) {
        List<MovieShow> shows = movieShowRepository.findAllByMovId(movie.getMovId());
        for (MovieShow show : shows) {
            log.info("Show at: " + show.getDate());
        }
        return shows;
    }

    public List<MovieShow> getShowsForMovieDate(Movie movie, String dateString) {
        return getShowsForMovieDate((int) movie.getMovId(), dateString);
    }

    public List<MovieShow> getShowsForMovieDate(Integer movId, String dateString) {
        Date date = DateUtils.createDateFromDateString(dateString);
        Date[] betweenDates = getBetweenDates(date);
        List<MovieShow> shows = movieShowRepository.findAllByMovIdAndDateBetweenOrderByDate(movId, betweenDates[0], betweenDates[1]);
        for (MovieShow show : shows) {
            log.info("Show at: " + show.getDate());
        }
        return shows;
    }

    public void saveMovieShow(Long movieId, Date date) {
        MovieShow show = new MovieShow();
        show.setMovId(movieId);
        show.setDate(date);
        log.info("Saving MovieShow: movId='{}' date='{}'", movieId, date);
        movieShowRepository.save(show);
    }

    public MovieShow saveMovieShowIfNotExists(Long movieId, Date date) {
        MovieShow show = new MovieShow();
        show.setMovId(movieId);
        show.setDate(date);
        if(!movieShowRepository.findAllByDateAndMovId(date,movieId).isEmpty()){
            log.info("MovieShow movId='{}' date='{}' already exists", movieId, date);
            return null;
        }
        log.info("Saving MovieShow: movId='{}' date='{}'", movieId, date);
        return movieShowRepository.save(show);
    }

    public void deleteMovieShow(Long showid) {
        log.info("Deleting MovieShow by showId='{}'", showid);
        movieShowRepository.deleteById(showid);
    }

    public void deleteAll() {
        log.info("Deleting all MovieShows");
        movieShowRepository.deleteAll();
    }

    public List<MovieShow> getShowsForDate(Date date, boolean groupByMovId) {
        log.info("Querying MovieShows for day: '{}'", date);
        Date[] betweenDates = getBetweenDates(date);
        return movieShowRepository.findAllByDateBetweenOrderByDate(betweenDates[0], betweenDates[1]);
    }

    /**
     * Return dates from today 00:00 and 23:59
     */
    private Date[] getBetweenDates(Date date) {
        Date dateStart = new Date(date.getTime());
        dateStart.setHours(0);
        dateStart.setMinutes(0);
        dateStart.setSeconds(0);
        Date dateEnd = new Date(date.getTime());
        dateEnd.setHours(23);
        dateEnd.setMinutes(59);
        dateEnd.setSeconds(59);
        return new Date[]{dateStart, dateEnd};
    }

    public MovieShow saveShow(MovieShow show) {
        log.info("Saving MovieShow for movId: '{}' day: '{}'", show.getMovId(), show.getDate());
        return movieShowRepository.save(show);
    }

    public Optional<MovieShow> queryShow(Long id) {
        log.info("Querying MovieShows by id: '{}'", id);
        return movieShowRepository.findById(id);
    }

    public void deleteById(Long id) {
        log.info("Deleting MovieShows by id: '{}'", id);
        movieShowRepository.deleteById(id);
    }

    public List<MovieShow> getShowsForWeekOf(Date date) {
        date.setHours(0);
        date.setMinutes(0);
        date.setSeconds(0);
        Calendar in7Days = Calendar.getInstance();
        in7Days.setTime(date);
        in7Days.set(Calendar.HOUR_OF_DAY, 23);
        in7Days.set(Calendar.MINUTE, 59);
        in7Days.set(Calendar.SECOND, 59);
        in7Days.add(Calendar.DATE, 7);
        log.info("Querying MovieShows between: '{}' and '{}'", date, in7Days.getTime());
        return movieShowRepository.findAllByDateBetweenOrderByDate(date, in7Days.getTime());
    }

    public List<MovieShow> getMovieShowsForWeekOf(Movie movie, Date date) {
        date.setHours(0);
        date.setMinutes(0);
        date.setSeconds(0);
        Calendar in7Days = Calendar.getInstance();
        in7Days.setTime(date);
        in7Days.set(Calendar.HOUR_OF_DAY, 23);
        in7Days.set(Calendar.MINUTE, 59);
        in7Days.set(Calendar.SECOND, 59);
        in7Days.add(Calendar.DATE, 7);
        log.info("Querying MovieShows for Movie '{}' for week of '{}'", movie.getName(), date);
        return movieShowRepository.findAllByMovIdAndDateBetweenOrderByDate(movie.getMovId(), date, in7Days.getTime());
    }

    public void saveShows(List<MovieShow> movieShows) {
        log.info("Saving MovieShows '{}'", Utils.joinToStringList(movieShows.stream().map(MovieShow::getShowId)));
        movieShows.stream().sorted(Comparator.comparing(MovieShow::getDate)).forEach(this::saveShow);
    }

    public List<MovieShow> getShowsForBetween(Date from, Date until) {
        log.info("Querying MovieShows between: '{}' and '{}'", from, until);
        return movieShowRepository.findAllByDateBetweenOrderByDate(from, until);
    }

    public List<MovieShow> getShowsForMovIdListAndBetweenDates(List<Long> movIds, Date from, Date until) {
        log.info("Querying MovieShows for Movies movIds='{}' between: '{}' and '{}'", Utils.joinToStringList(movIds), from, until);
        return movieShowRepository.findAllByMovIdInAndDateBetweenOrderByDate(movIds, from, until);
    }

    public long deleteShowsByIds(List<Long> showIds) {
        log.info("Deleting MovieShows by ids='{}'", Utils.joinToStringList(showIds));
        return movieShowRepository.deleteAllByShowIdIn(showIds);
    }


}

File: movieland-backend/rest/service/movie_show_service.py
from datetime import datetime, timedelta
from typing import List

from db.database import db
from db.model import MovieShow
from logger import get_logger
from rest.service.reservation_service import ReservationService

log = get_logger()

res_service= ReservationService()

class MovieShowService:

    def get_show_by_id(self, id: int) -> MovieShow:
        return MovieShow.query.filter_by(showId=id).first()

    def get_shows_for_date(self, date: datetime, movId: int = None) -> List[MovieShow]:
        log.info(f"Querying MovieShows for date='{date}' and movIds='{movId}' ")
        date_from = date.replace(hour=0, minute=0, second=0)
        date_until = date.replace(hour=23, minute=59, second=59)
        if movId is not None:
            return MovieShow.query.filter(MovieShow.date >= date_from, MovieShow.date <= date_until,
                                          MovieShow.movId == movId).all()
        return self.get_shows_for_between(date_from, date_until)

    def get_shows_of_week(self, date: datetime = datetime.today()) -> List[MovieShow]:
        date_from = date.replace(hour=0, minute=0, second=0)
        date_until = (date_from + timedelta(days=7)).replace(hour=23, minute=59, second=59)
        log.info(f"Querying MovieShows between: '{date_from}' and '{date_until}'")
        return self.get_shows_for_between(date_from, date_until)

    def get_shows_of_week_of_movie(self, movId: int, date: datetime = datetime.today()) -> List[MovieShow]:
        start_datetime = date.replace(hour=0, minute=0, second=0)
        end_datetime = (start_datetime + timedelta(days=7)).replace(hour=23, minute=59, second=59)
        log.info(f"Querying MovieShows between: '{start_datetime}' and '{end_datetime}'")
        return MovieShow.query.filter(MovieShow.date >= start_datetime, MovieShow.date <= end_datetime,
                                      MovieShow.movId == movId).all()

    def get_shows_for_between(self, date_from, date_until) -> List[MovieShow]:
        _from = date_from.replace(hour=0, minute=0, second=0)
        _until = date_until.replace(hour=23, minute=59, second=59)
        return MovieShow.query.filter(MovieShow.date >= _from, MovieShow.date <= _until).all()

    def save_show(self, show: MovieShow, update: bool = False):
        if update:
            db.session.merge(show)
        else:
            db.session.add(show)
        db.session.commit()
        return show

    def delete_all(self):
        MovieShow.query.delete()
        db.session.commit()

    def delete_by_ids(self, show_ids: List[int]):
        res_service.delete_reservations_of_shows(show_ids)
        MovieShow.query.filter(MovieShow.showId.in_(show_ids)).delete()
        db.session.commit()

    def save_shows(self, movie_shows: List[MovieShow]):
        for show in movie_shows:
            self.save_show(show)

    def get_shows_for_mov_ids_and_between(self, mov_ids: List[int], date_from: datetime, date_until: datetime) -> List[MovieShow]:
        return MovieShow.query.filter(MovieShow.movId.in_(mov_ids), MovieShow.date >= date_from,
                                      MovieShow.date <= date_until).all()

    def get_by_movies(self, mov_ids: List[int]) -> List[MovieShow]:
        return MovieShow.query.filter(MovieShow.movId.in_(mov_ids)).all()
